### PR TITLE
add github ci

### PR DESCRIPTION
**Pull request by nmm0**
Currently this is pretty bare-bones, but eventually we will expand our testing configurations.

### CODE REVIEW

**Review by keitat — DISMISSED** [review]

File: CMakeLists.txt [hunk 1]
```diff
@@ -101,6 +101,19 @@ if (KR_ENABLE_VELOC OR KR_ENABLE_HDF5_PARALLEL)
    target_link_libraries(resilience PRIVATE MPI::MPI_CXX)
 endif()
 
+option(KR_ENABLE_EXEC_SPACES "enable resilient execution spaces" OFF)
```
Comment (nmm0) [inline CMakeLists.txt:104]: Added this flag to gate all resilient exec space stuff

File: ci/spack.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
+spack:
```
Comment (nmm0) [inline ci/spack.yaml:1]: Here we are using a spack environment to set up a build for all dependencies (except for veloc)

File: ci/ubuntu20.04-gcc11-x64.dockerfile [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  && spack env activate . \
+  && spack concretize -f
+
+RUN cd /opt/spack-environment \
+  && . /opt/spack/share/spack/setup-env.sh \
+  && spack env activate . \
+  && spack install --fail-fast \
+  && spack gc -y
+
+# We need to build a specific branch of VeloC until https://github.com/ECP-VeloC/VELOC/pull/43 is resolved
```
Comment (nmm0) [inline ci/ubuntu20.04-gcc11-x64.dockerfile:61]: I have a separate build for veloc, even though 43 was merged in (just now), the spack package is kind of outdated and spack updates their packages slowly. So we will still need this.

File: src/resilience/Config.hpp [hunk 4]
```diff
@@ -82,7 +82,12 @@ namespace KokkosResilience
       template< typename T, typename = std::enable_if_t< std::is_constructible< variant_type, T >::value > >
       explicit Value( T &&val )
       {
-        m_variant = std::forward< T >( val );
+        if constexpr ( std::is_constructible_v< std::string, std::add_rvalue_reference_t< T > > )
```
Comment (nmm0) [inline src/resilience/Config.hpp:85]: I had to driveby fix a bug where `const char *` ptrs where being auto converted to bool with gcc 11 (this is not the case for gcc 12)

File: CMakeLists.txt (hunk 1 above)
Comment (Matthew-Whitlock) [reply CMakeLists.txt:104]: This is minor, but would something like `cmake_dependent_option(KR_ENABLE_CUDA_EXEC_SPACE "enable resilient cuda execution space" ON "KR_ENABLE_EXEC_SPACES;_cuda_found GREATER -1" OFF)` work?

**Review by Matthew-Whitlock — APPROVED** [review]
I'm not familiar enough with Spack to give a good review there, but it looks reasonable and it works.

**Comment by nmm0** [comment]
@Matthew-Whitlock good idea, I will try that
